### ISSUE TITLE
HTTP2: Race in processing trailing headers

### ISSUE THREAD

**Issue by geoffkizer** [dotnet/corefx#39427]
For HTTP2, we are processing trailing headers in Http2Stream.OnResponseHeader. This is called when we receive a HEADERS frame. We directly add any trailing headers to the associated HttpResponseMessage's TrailingHeaders collection. This means we're modifying the TrailingHeaders collection while another thread may be examining it. 

In #34912, we agreed that trailing headers would not be added to the TrailingHeaders collection until the user reads to the end of the response stream. This avoids the race and makes TrailingHeaders availability deterministic. This is indeed how HTTP/1.1 works; however we unfortunately did not do this for HTTP2.

**Comment by geoffkizer** [comment]
In practice, I think we will need to store the trailing headers in a temporary List<> or similar and then only add them to the TrailingHeaders collection when the user reads to end of stream. This isn't ideal, but I don't see any better way to handle this. TrailingHeaders are rare enough that I don't think it's worth worrying about.

**Comment by stephentoub** [comment]
I'm confused by the issue. Are you saying we add to the trailing headers after the steam returns EOF? We said that no one should read from TrailingHeaders until then, not that we would delay adding headers until then (though the latter would be an acceptable implementation).

**Comment by geoffkizer** [comment]
> Are you saying we add to the trailing headers after the stream returns EOF? We said that no one should read from TrailingHeaders until then, not that we would delay adding headers until then (though the latter would be an acceptable implementation)

That's my understanding of what we decided, yes. Specifically that TrailingHeaders would not be available until the stream returns EOF. See:

https://github.com/dotnet/corefx/issues/34912#issuecomment-461229594
> The TrailingHeaders will be available only after the underneath response body stream has reached EOF.

We did this for several reasons:
(1) Avoid non-deterministic behavior with TrailingHeaders. We wanted to avoid a situation where the user doesn't read to EOF, but sometimes the TrailingHeaders are available anyway, and sometimes they are not. 
(2) Avoid the sort of race described here. That is, if the user accesses TrailingHeaders before receiving EOF. We wanted to ensure that it's always safe to access anything on the HttpResponseMessage, including TrailingHeaders, as long as you are not using the HttpResponseMessage from multiple threads at the same time -- where "using the HttpResponseMessage" is defined to include reading from the response stream.

There's some discussion of this in the linked issue, though it's buried among a bunch of other issues as well.

It turns out that this also helps with the rooting issue, but I don't think we considered that at the time.
